### DISCUSSION TITLE
Non-unique metrics exposed on per-object metrics endpoint

### DISCUSSION THREAD

**Discussion by vCra** [rabbitmq/rabbitmq-server#9665] [Other]
### Describe the bug

When scraping the /metrics/per-object endpoint, we occasionally see duplicated metrics returned from RMQ

These metrics sometimes have the correct labels, and sometimes do not. This issue doesn't affect the entire cluster - it's usually just one node

This is causing issues with ingestion into prometheus, as it sees these identical metrics as out-of-order, which generated errors

Metrics which seem to be affected

* rabbitmq_raft_term_total
* rabbitmq_raft_log_snapshot_index
* rabbitmq_raft_log_last_applied_index
* rabbitmq_raft_log_commit_index
* rabbitmq_raft_log_last_written_index
* rabbitmq_raft_entry_commit_latency_seconds

### Reproduction steps

We are struggling to reproduce this

### Expected behavior

All the metrics should have unique labels

**Comment by michaelklishin** [comment]
@vCra how do you define "an incorrect label"? I can be a race condition between the table that stores individual metric values, e.g. it does not necessarily require the keys to be unique.

But that would not result in any "incorrect" labels.

**Reply by vCra** [reply]
Hi @michaelklishin

incorrect is probably the wrong phrasing here - we're sometimes seeing metrics without any labels at all - see the following example

<img width="978" alt="Screenshot 2023-10-16 at 13 25 53" src="https://github.com/rabbitmq/rabbitmq-server/assets/22872863/176bbbea-6432-4209-b0c0-bd7a995a4175">

The expected behaviour would be for all of these metrics to have labels

**Reply by michaelklishin** [reply]
Not all Ra features (and thus metrics) are associated with a quorum queue or stream. Raft was used for MQTT client ID tracking up to a point. Does this system use or have the MQTT plugin enabled? My best guess is that some of those metrics are not associated with a queue.

At the same time, there should not be that many of them :/

**Reply by michaelklishin** [reply]
Another hypothesis we have: can those queues be gone by now but remain in the metrics table? Do you experience much quorum queue churn in your system?

**Reply by vCra** [reply]
We do have an MQTT listener enabled, but we don't use it at all - we just use AMQP
We don't automatically churn queues or anything - maybe manually remove 1-2 a year at most

**Reply by dbereza-machinify** [reply]
Hello, was wondering if there are any updates on this?

We are also seeing a similar issue crop up on RabbitMQ 3.13.6 for the `/metrics` endpoint where duplicate metrics are being outputted, the same six listed in the initial post: 

rabbitmq_raft_term_total
rabbitmq_raft_log_snapshot_index
rabbitmq_raft_log_last_applied_index
rabbitmq_raft_log_commit_index
rabbitmq_raft_log_last_written_index
rabbitmq_raft_entry_commit_latency_seconds

Example duplicate output:
```
$ curl http://172.60.18.168:15692/metrics | grep "rabbitmq_raft_term_total "
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# TYPE rabbitmq_raft_term_total counter
# HELP rabbitmq_raft_term_total Current Raft term number
rabbitmq_raft_term_total 259
rabbitmq_raft_term_total 40
```

Similarly, we don't have the MQTT plugin enabled nor have any real queue churn, although we do pretty much exclusively use quorum and stream queues.

**Reply by michaelklishin** [reply]
There are no updates except for one: we know when this duplicate metric can be returned by the HTTP API: when a quorum queue is in the process of transferring leadership, so for a short period of time both nodes will have local leader metrics.

I don't see how this can affect the Prometheus `/metrics` endpoint which returns data from a single node only.

Those who are interested in getting to the bottom of this are welcome to investigate. This is open source software after all.

**Reply by vCra** [reply]
We just ended up dropping raft related metrics whilst scraping, as we didn't have any need for them

```hcl
          "metricRelabelings" = [
            {
              "sourceLabels" = [
                "__name__"
              ]
              "action" = "drop"
              "regex"  = "rabbitmq_raft_term_total|rabbitmq_raft_log_snapshot_index|rabbitmq_raft_log_last_applied_index|rabbitmq_raft_log_commit_index|rabbitmq_raft_log_last_written_index|rabbitmq_raft_entry_commit_latency_seconds"
            }
          ]
```
